### PR TITLE
addTest/updateProject.spec.js/issue#570

### LINKED ISSUE #570: Resolvers: Create tests for updateProject.js
The Talawa-API code base needs to be 100% reliable. This means we need to have 100% test code coverage.

Tests need to be written for file `lib/resolvers/event_project_mutations/updateProject.js`

- We will need the API to be refactored for all methods, classes and/or functions found in this file for testing to be correctly executed.
- When complete, all all methods, classes and/or functions in the refactored file will need to be tested. These tests must be placed in a
single file with the name `tests/resolvers/event_project_mutations/updateProject.spec.js`. You may need to create the appropriate directory structure to do this.

### IMPORTANT:
Please refer to the parent issue on how to implement these tests correctly:

- https://github.com/PalisadoesFoundation/talawa-api/issues/490

### PR Acceptance Criteria

- When complete this file must show **100%** coverage when merged into the code base. This will be clearly visible when you submit your PR.
- [The current code coverage for the file can be found here](https://codecov.io/gh/PalisadoesFoundation/talawa-api/tree/develop/lib/resolvers/event_project_mutations/). If the file isn't found in this directory, or there is a 404 error, then tests have not been created.
- The PR will show a report for the code coverage for the file you have added. You can use that as a guide.

### PR DESCRIPTION
**Pull request by xoldyckk**
**What kind of change does this PR introduce?**

Refactors updateProject.js.
Adds test file for updateProject.js

**Issue Number:**

Fixes #570

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

Screenshot from 2022-03-26 11-02-42

**If relevant, did you update the documentation?**

No

**Summary**

Refactored some files to throw easily mockable errors when not in production environment. Fixed a bug where comparing mongodb's objectId was implemented incorrectly. Added test file for updateProject.js increasing code coverage across talawa-api project.

**Does this PR introduce a breaking change?**

No

**Other information**

No

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
Can you get the code coverage up to 100%?

**Comment by xoldyckk** [comment]
> Can you get the code coverage up to 100%?

The errors thrown by default in resolvers are not mockable. When mocked they throw `i18n is not initialized, try app.use(i18n.init)` as the error message. This occurs because of the function `requestContext.translate`. This function only runs inside the context of the actual talawa-api server. That means we have to simulate an actual talawa-api server inside our tests to test it. That's why there's this refactor :- 

      process.env.NODE_ENV !== 'production'
        ? 'User not authorized'
        : requestContext.translate('user.notAuthorized')

So, that in testing environment the errors don't call requestContext.translate function. Instead they use a plain string.

And I believe that's what's causing the test coverage to not be 100%.

**Comment by palisadoes** [comment]
Can we mock `requestContext.translate` in some way?

Should we consider running an API server as part of the testing regimen and mock where necessary for the rest? I've written code in the past where I had to spin up an instance of the API to properly test end to end. I realize that this may not be practical in future where running the API may consume all test resources.

**Comment by xoldyckk** [comment]
> Can we mock `requestContext.translate` in some way?
> 
> Should we consider running an API server as part of the testing regimen and mock where necessary for the rest? I've written code in the past where I had to spin up an instance of the API to properly test end to end. I realize that this may not be practical in future where running the API may consume all test resources.


I personally don't have much experience with testing. I only got recently got started with it because of this project. So, my suggestions could be wrong or unreliable. But here's my take:-

In all the integration/e2e test implementations I've seen on testing backends/server, instead of just testing the route/resolver in isolation, they all run tests on the actual server. To make testing in CI runs faster, we can pass the option `--runInBand` to the test script to make it faster. Here's the source :- [https://dev.to/vantanev/make-your-jest-tests-up-to-20-faster-by-changing-a-single-setting-i36](url).

**Comment by ShwetKhatri2001** [comment]
I'm also getting the same problem of not covering the second branch ( involving requestContext.translate ) in my open PRs.
I tried to mock it by initiating i18n with the same settings we have in server file ( `index.js` ) , but didn't get expected behaviour.

But according to me here the scope of our API test case is to check if the error is delivered, not in which language it is delivered. That's why we separated

**Comment by palisadoes** [comment]
@JamaicanFriedChicken @DMills27 This was the case I spoke about today. Please please give your inputs.

**Comment by palisadoes** [comment]
@JamaicanFriedChicken @DMills27 Please respond here so we have a record of your recommended approach.

**Comment by JamaicanFriedChicken** [comment]
@xoldyckk @ShwetKhatri2001 Thanks guys. We will be making this into a separate issue referencing this PR to investigate the cause of this. For now, we will accept the PRs as is, as it relates to i18n even if the test coverage isn't 100%.

File: tests/resolvers/event_project_mutations/updateProject.spec.js [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+let testUser1;
+
+let testUser2;
+
+let testOrganization;
+
+let testEvent;
+
+let testEventProject;
```
Comment (JamaicanFriedChicken) [inline tests/resolvers/event_project_mutations/updateProject.spec.js:123]: it's best to put these variables at the top of the file.

**Review by JamaicanFriedChicken — CHANGES_REQUESTED** [review]
Still need to make some slight adjustments, almost there.

**Review by JamaicanFriedChicken — APPROVED** [review]

**Pull request merged**